### ISSUE TITLE
Unexpected behaviour of hooking inside callback

### ISSUE THREAD

**Issue by SpirinEgor** [PyTorchLightning/pytorch-lightning#5917]
## 🐛 Bug

`on_epoch_end` hook in callback called between train and validation epochs.

### To Reproduce

Use following callback
```(python)
class BuggyCallback(Callback):
    def on_epoch_end(self, trainer: Trainer, pl_module: LightningModule):
        print("I'm called")
```

### Expected behavior

Line with `I'm called` should appear after train and validation cycles. Instead of this, that line printed exactly after the training cycle, before validation.

### Environment

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux): MacOS BigSur
 - How you installed PyTorch (`conda`, `pip`, source): conda
 - Build command you used (if compiling from source): -
 - Python version: 3.8.5
 - CUDA/cuDNN version: CPU version
 - GPU models and configuration: -
 - Any other relevant information: -

### Additional context

I checked manually that it works properly in the 1.1.7 version. Thus, the bug appears only in the last version of lightning.

**Comment by tchaton** [comment]
Hey @SpirinEgor,

Great catch there !

Best,
T.C

**Comment by egonuel** [comment]
Hi, I just updated to 1.2 and had skipped some pl updates prior to this (must have been <= 1.1.7 as per reporters comment), but now I have some issues with my reporting callbacks, which appear to be related to this.
As far as I remember the `on_epoch_end()` was called after the training step and the validation loop per epoch, so that this was the point where one had access to training and validation losses and metrics for a single epoch.
Calling order must have been like so (with everything else not shown):
```
on_train_epoch_end()
on_validation_epoch_end()
on_epoch_end()
```
Now, it looks like so:
```
on_train_epoch_end()
on_epoch_end()
on_validation_epoch_end()
on_epoch_end()
```
Is this the intended behavior as there is no entry point for the "ultimate" epoch end anymore?
In addition, `on_epoch_end()` is also called after the sanity validation steps, which, in my understanding, should not happen.
`on_epoch_start()` is BTW still the "ultimate" start of the epoch, prior to everything.

**Comment by addisonklinke** [comment]
@tchaton I think this should be re-opened. I just tested in v1.2.1 and notice the same unexpected order as @egonuel's second code block. I strongly believe the hook order should follow his first code block, so that we can have an entry point for the "ultimate" epoch end as mentioned.

Also, in the list of available hooks from the [docs](https://pytorch-lightning.readthedocs.io/en/stable/extensions/callbacks.html#available-callback-hooks) we have

```
on_validation_start
on_validation_epoch_start
on_validation_epoch_end
on_validation_end

on_train_start
on_train_epoch_start
on_train_epoch_end
on_train_end
```

I can't find any useful clarification of the difference between `on_train/validation_start/end` vs. the `_epoch_` variation and when users should prefer one over the other. Could you please clarify? Perhaps we can create a PR to put the complete order of hooks into the docs (as a list and as pseudo-code). I would find that very helpful as a reference of expected behavior.

**Comment by SpirinEgor** [comment]
Completely agreed with @egonuel that using several `on_epoch_end()` during one epoch is discourage. The epoch ends only ones, so `on_epoch_end()` should be called only ones.

As far as I remember, in previous versions order of hooks were:
```
on_epoch_start
on_train_epoch_start
on_validation_epoch_start
on_validation_epoch_end
on_train_epoch_end
on_epoch_end
```
Which is quietly straightforward. And here `on_epoch_end()` can be used to aggregate information from the completed epoch, e.g. printing some metrics or uploading checkpoints.

I suggest reopening this issue for further discussion :) @tchaton what do you think?

**Comment by addisonklinke** [comment]
@SpirinEgor I notice @williamFalcon mentions on a [couple](https://github.com/PyTorchLightning/pytorch-lightning/issues/3832#issuecomment-703200707) of other issue [threads](https://github.com/PyTorchLightning/pytorch-lightning/issues/4001#issuecomment-705844942) that users commonly forget the possibility of multiple validation runs per training epoch. This is a good point, so we'd probably want to update the desired order of hooks as

```
on_epoch_start
on_train_epoch_start

on_validation_epoch_start
on_validation_epoch_end

on_validation_epoch_start
on_validation_epoch_end

... N validation repeats ...

on_train_epoch_end
on_epoch_end
```

As I understand the current behavior (desired or not is unclear to me), `on_epoch_end` is called every time `on_*_epoch_end` is called. In this way it seems to be a duplicate, albeit less specific signal that allows you to hook into the end of any type of epoch whether that's train, validation, or test. 

What I believe the users on this thread are looking for is a signal for the end of an _Epoch_ (capital "E") that consists of 1 training epoch (lowercase "e") and one or more validation epoch(s). To me, the concept of an Epoch generalizes regardless of whether you're training on a huge dataset (where multiple validation runs are desirable) or a small dataset (where a single validation run is sufficient). 

The current `on_train_epoch_end` seems close, but at least in my case it doesn't have the aggregated validation metrics available automatically in `trainer.callback_metrics`. For now, `on_validation_epoch_end` works if you only have one validation run per training epoch. However, if you want multiple validation runs, I am struggling to find a good way to summarize them once the training epoch completes. Maybe you'd need a stateful custom callback that hooks `on_validation_epoch_end` and keeps a memory of all N validation runs. Then when `on_train_epoch_end` is triggered, consume all the validation stats and flush them during the next `on_train_epoch_start`?

**Comment by SpirinEgor** [comment]
Yes, it is a good point about multiple validations during one epoch. I agree that the order should look like you mention.
I'm already using the custom hook similar to the one you described. But I find it as a temporary solution.

I think the case with multiple validations can be handled with a simple list of results inside the trainer.

**Comment by jbohnslav** [comment]
I agree that this should be reopened. I can't upgrade to 1.2 because of this. I have code that is run when validation ends, e.g. computing metrics over the validation set. This logs metrics for use in learning rate scheduling, etc. Because `on_epoch_end` is now called before `on_validation_epoch_ends`, none of my metrics exist, and I get a `pytorch_lightning.utilities.exceptions.MisconfigurationException: ReduceLROnPlateau conditioned on metric my_metric which is not available.`

Downgrading to 1.1.8 fixes this issue.

**Comment by egonuel** [comment]
@jbohnslav I don't know if this is useful for you, but I hacked a temporary solution for my callbacks like this: 
```
class TempFixedOnEpochEndCallback(Callback):
    def __init__(self):
        self.is_training = False

    def on_train_start(self, trainer, pl_module) -> None:
        self.is_training = True

    def on_train_end(self, trainer, pl_module) -> None:
        self.is_training = False

    def on_train_epoch_end(self, trainer, pl_module, outputs) -> None:
        if trainer.val_dataloaders is None:
            self.do_something(trainer)

    def on_validation_epoch_end(self, trainer, pl_module) -> None:
        if self.is_training:
            self.do_something(trainer)
            
    def doe_something(self, trainer):
        pass
```
At least under my conditions, this allows me to get the old behavior back. This also prevents the call in the sanity checks. No guarantee that this works everywhere, though 🙂
